### PR TITLE
Rewrite LinearRegressor

### PR DESCRIPTION
**Pull request by snnn**
**Description**: 

Rewrite LinearRegressor so that it can get benefits from mlas when data size is large

**Motivation and Context**
- Why is this change required? What problem does it solve?

onnxruntime is not performant on batch prediction on traditional ml models.

- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/core/providers/cpu/ml/linearregressor.cc [hunk 1]
```diff
@@ -1,61 +1,48 @@
...
-    return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT,
-                  "Input shape needs to be at least a single dimension.");
+  c.clear();
+  ORT_ENFORCE(info.GetAttrs<float>("coefficients", c).IsOK());
+  coefficients_.resize(c.size());
+  ORT_ENFORCE(c.size() % targets_ == 0);
+  int64_t feature_size = static_cast<int64_t>(c.size()) / targets_;
+  for (int64_t i = 0; i != targets_; ++i) {
+    int64_t offset = i * feature_size;
+    for (int64_t j = 0; j != feature_size; ++j) {
```
Comment (xadupre) [inline onnxruntime/core/providers/cpu/ml/linearregressor.cc]: Maybe the use of pointers would help here, at least for c, for coefficients_ too if targets_ is 1.

File: onnxruntime/core/providers/cpu/ml/linearregressor.h [hunk 2]
```diff
@@ -15,12 +16,50 @@ template <typename T>
...
+    Tensor* Y = ctx->Output(0, TensorShape({N, targets_}));
+    T* Ydata = Y->MutableData<T>();
+    math::MatMul<T>(static_cast<int>(N), static_cast<int>(targets_), static_cast<int>(stride), X->Data<T>(),
+                    coefficients_.data(), Y->MutableData<T>(), tp);
+    bool useIntercepts = intercepts_.size() == static_cast<size_t>(targets_);
+    if (useIntercepts) {
+      for (int64_t i = 0; i < N; i++)  // for each point
+      {
+        T* p = Ydata + i * targets_;
+        for (int j = 0; j < targets_; j++)  // for each target
```
Comment (xadupre) [inline onnxruntime/core/providers/cpu/ml/linearregressor.h]: Same comment, probably faster with pointers.

File: onnxruntime/core/providers/cpu/ml/ml_common.h [hunk 3]
```diff
@@ -234,125 +195,111 @@ static inline void multiclass_probability(int64_t classcount, const std::vector<
...
+// It assumes val is in [0,1]
+template <typename T>
+inline T ComputeProbit(T val) {
+  return static_cast<T>(M_SQRT2) * ErfInv(2 * val - 1);
 }
 
 static inline float sigmoid_probability(float score, float proba, float probb) {
   float val = score * proba + probb;
-  return 1 - ComputeLogistic(val);  // ref: https://github.com/arnaudsj/libsvm/blob/eaaefac5ebd32d0e07902e1ae740e038eaaf0826/svm.cpp#L1818
+  return 1 -
```
Comment (xadupre) [inline onnxruntime/core/providers/cpu/ml/ml_common.h]: I'm curious, you folded enums in one line but you unfolded this line into 3.

File: onnxruntime/core/providers/cpu/ml/ml_common.h [hunk 4]
```diff
@@ -234,125 +195,111 @@ static inline void multiclass_probability(int64_t classcount, const std::vector<
...
-            scores.push_back(ComputeLogistic(scores[0]));  //ml_logit(scores[k]);
-            scores[0] = ComputeLogistic(-scores[0]);
-          } else {
-            scores.push_back(-scores[0]);
-          }
-          break;
+void write_scores(T* scores, size_t scores_len, POST_EVAL_TRANSFORM post_transform, T* out_p) {
+  switch (post_transform) {
+    case POST_EVAL_TRANSFORM::PROBIT:
+      for (size_t i = 0; i != scores_len; ++i) out_p[i] = ComputeProbit(scores[i]);
```
Comment (xadupre) [inline onnxruntime/core/providers/cpu/ml/ml_common.h]: Pointers?

File: onnxruntime/core/util/softmax.h [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        (softmax - softmax.exp().sum(along_class).log().eval().reshape(batch_by_one).broadcast(one_by_class));
+  } else {
+    // NOTE(touts): If you modify this implementation please run
+    // the BM_ImageNetSoftmaxFwd benchmark in nn_ops_test.cc.
+    //
+    // softmax = exp(logits - max(logits along classes));
+    softmax.device(d) = shifted_logits.exp();
+    // softmax = softmax * (1 / sum(softmax along classes));
+    softmax.device(d) =
+        (softmax * softmax.sum(along_class).inverse().eval().reshape(batch_by_one).broadcast(one_by_class));
```
Comment (xadupre) [inline onnxruntime/core/util/softmax.h]: What happens if one value is 0 or if you have 0 / 0 ?

File: onnxruntime/core/providers/cpu/ml/linearregressor.cc (hunk 1 above)
Comment (snnn) [reply onnxruntime/core/providers/cpu/ml/linearregressor.cc]: Thanks, will try.

Comment (snnn) [reply onnxruntime/core/providers/cpu/ml/linearregressor.cc]: Hi , I wrote a small test, and it didn't show any perf difference.  Pointer is not faster.

```C++
static void BM_Transpose1(benchmark::State& state) {
  int targets_ = 1024;
  std::vector<float> c(1024 * 10);
  std::vector<float> coefficients_(1024 * 10);
  std::uniform_real_distribution<float> d(-1, 1);
  for (size_t i = 0; i != coefficients_.size(); ++i) {
    c[i] = d(mt);
  }
  ORT_ENFORCE(c.size() % targets_ == 0);
  for (auto _ : state) {
    int64_t feature_size = static_cast<int64_t>(c.size()) / targets_;
    for (int64_t i = 0; i != targets_; ++i) {
      int64_t offset = i * feature_size;
      for (int64_t j = 0; j != feature_size; ++j) {
        coefficients_[j * targets_ + i] = c[offset + j];
      }
    }
  }
}

static void BM_Transpose2(benchmark::State& state) {
  int targets_ = 1024;
  std::vector<float> c(1024 * 10);
  std::vector<float> coefficients_(1024 * 10);
  std::uniform_real_distribution<float> d(-1, 1);
  for (size_t i = 0; i != coefficients_.size(); ++i) {
    c[i] = d(mt);
  }
  ORT_ENFORCE(c.size() % targets_ == 0);
  for (auto _ : state) {
    int64_t feature_size = static_cast<int64_t>(c.size()) / targets_;
    for (int64_t i = 0; i != targets_; ++i) {
      float* p1 = coefficients_.data() + i;
      float* p2 = c.data() + i * feature_size;
      for (int64_t j = 0; j != feature_size; ++j) {
        p1[j * targets_] = p2[j];
      }
    }
  }
}
BENCHMARK(BM_Transpose1);
BENCHMARK(BM_Transpose2);
```

File: onnxruntime/core/providers/cpu/ml/ml_common.h (hunk 3 above)
Comment (snnn) [reply onnxruntime/core/providers/cpu/ml/ml_common.h]: The original code is too long, longer than 120 chars.

File: onnxruntime/core/providers/cpu/ml/ml_common.h [hunk 6]
```diff
@@ -234,125 +219,109 @@ static inline void multiclass_probability(int64_t classcount, const std::vector<
...
-    values[k] = newscores[k] / this_sum;
-  }
+/**
+ * if scores.size() == 1 and post_transform != POST_EVAL_TRANSFORM::PROBIT and
+ * add_second_class is in [0, 3], output 2 value else output scores.size()
+ * values
+ */
+template <typename T>
+void write_scores(std::vector<T>& scores, POST_EVAL_TRANSFORM post_transform, int64_t write_index, Tensor* Z,
+                  int add_second_class) {
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/ml/ml_common.h]: Are there perf gains from the changes to this method as the new implementation is a lot harder to follow?

File: onnxruntime/core/providers/cpu/ml/linearregressor.cc [hunk 7]
```diff
@@ -1,61 +1,59 @@
...
-    yindex += scores.size();
+    std::vector<float> c = info.GetAttrsOrDefault<float>("intercepts");
+    intercepts_.resize(c.size());
+    std::copy_n(c.data(), c.size(), intercepts_.data());
+  }
+
+  // A dirty hack to keep the code working as before
+  if (targets_ == 1) {
+    // In RS4, we implemented the PROBIT transform for single class cases,
+    // but the outputted value is most likely NaN
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/ml/linearregressor.cc]: Why should the ORT code have something that changes the behavior based on things that happened in RS4? That seems like something that should be somewhere other than master.

Comment (snnn) [reply onnxruntime/core/providers/cpu/ml/linearregressor.cc]: Still output NaN randomly?

File: onnxruntime/core/providers/cpu/ml/linearregressor.h [hunk 8]
```diff
@@ -15,12 +16,50 @@ template <typename T>
...
+                    "Input shape needs to be at least a single dimension.");
+    }
+    // X: [N, feature_size],coefficients_[target, feature_size]
+    // X*coefficients_^t : [N, target]
+    int64_t stride = X->Shape().NumDimensions() == 1 ? X->Shape()[0] : X->Shape()[1];
+    int64_t N = X->Shape().NumDimensions() == 1 ? 1 : X->Shape()[0];
+    Tensor* Y = ctx->Output(0, TensorShape({N, targets_}));
+    T* Ydata = Y->MutableData<T>();
+    //math::Gemm doesn't support double in all the platforms, so we use MatMul here.
+    math::MatMul<T>(static_cast<int>(N), static_cast<int>(targets_), static_cast<int>(stride), X->Data<T>(),
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/ml/linearregressor.h]: Would it be faster to use Gemm if T != double? Easily checked with std::is_same

File: onnxruntime/core/providers/cpu/ml/linearregressor.h [hunk 9]
```diff
@@ -15,12 +16,50 @@ template <typename T>
...
+                    coefficients_.data(), Y->MutableData<T>(), tp);
+    bool useIntercepts = intercepts_.size() == static_cast<size_t>(targets_);
+    if (useIntercepts) {
+      for (int64_t i = 0; i < N; i++)  // for each point
+      {
+        T* p = Ydata + i * targets_;
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/ml/linearregressor.h]: nit: run clang-format on this file as the brace layout is incorrect.

File: onnxruntime/core/providers/cpu/ml/linearregressor.h [hunk 10]
```diff
@@ -15,12 +16,50 @@ template <typename T>
...
+    int64_t N = X->Shape().NumDimensions() == 1 ? 1 : X->Shape()[0];
+    Tensor* Y = ctx->Output(0, TensorShape({N, targets_}));
+    T* Ydata = Y->MutableData<T>();
+    //math::Gemm doesn't support double in all the platforms, so we use MatMul here.
+    math::MatMul<T>(static_cast<int>(N), static_cast<int>(targets_), static_cast<int>(stride), X->Data<T>(),
+                    coefficients_.data(), Y->MutableData<T>(), tp);
+    bool useIntercepts = intercepts_.size() == static_cast<size_t>(targets_);
+    if (useIntercepts) {
+      for (int64_t i = 0; i < N; i++)  // for each point
+      {
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/ml/linearregressor.h]: Can you use ParallelFor for this or is it not worth it?

File: onnxruntime/core/providers/cpu/ml/linearregressor.h [hunk 11]
```diff
@@ -15,12 +16,50 @@ template <typename T>
...
+  }
 
  private:
   int64_t targets_;
-  std::vector<float> coefficients_;
-  std::vector<float> intercepts_;
+  Eigen::Tensor<T, 2, Eigen::RowMajor, Eigen::DenseIndex> coefficients_;
+  std::vector<T> intercepts_;
+  // Indeed, it can't be PROBIT
   POST_EVAL_TRANSFORM post_transform_;
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/ml/linearregressor.h]: I don't understand the comment

File: onnxruntime/core/providers/cpu/ml/linearregressor.h (hunk 10 above)
Comment (snnn) [reply onnxruntime/core/providers/cpu/ml/linearregressor.h]: Which ParallelFor ?  ParallelFor from the threadpool or eigen device?

File: onnxruntime/core/providers/cpu/ml/linearregressor.h (hunk 11 above)
Comment (snnn) [reply onnxruntime/core/providers/cpu/ml/linearregressor.h]: PROBIT requires the input be in [0,1] range. In this case, we can't guarantee that.

Comment (skottmckay) [reply onnxruntime/core/providers/cpu/ml/linearregressor.h]: Would be good to capture that info in the comment.

---
In reply to: [336697258](https://github.com/microsoft/onnxruntime/pull/1959#discussion_r336697258) [](ancestors = 336697258)

File: onnxruntime/core/providers/cpu/ml/linearregressor.h (hunk 10 above)
Comment (skottmckay) [reply onnxruntime/core/providers/cpu/ml/linearregressor.h]: I was thinking of the ThreadPool method but whatever works best. Not required for this PR.

---
In reply to: [336697105](https://github.com/microsoft/onnxruntime/pull/1959#discussion_r336697105) [](ancestors = 336697105)

File: onnxruntime/core/providers/cpu/ml/linearregressor.cc (hunk 7 above)
Comment (skottmckay) [reply onnxruntime/core/providers/cpu/ml/linearregressor.cc]: Just going off the ONNX spec, it seems like this approach is inconsistent as you're ignoring the post transform value completely if targets_ == 1. 
Is that because the spec should say that is the correct behavior, or just because previously we didn't handle this case (or handled it badly)? Or because we're making in implementation choice to not support this usage? If the latter, seems like we should fail or output a warning.

Similarly, just because there wasn't a PROBIT implementation previously for multi class cases, the ONNX spec says it's supported, so also turning off PROBIT if targets_ is != 1 seems like we don't support PROBIT anywhere. If that's the case shouldn't we at least fail the model (it's not going to do what a default user of ORT expects) or at a minimum output a warning message saying we're overriding the post transform value in the model?



---
In reply to: [336695370](https://github.com/microsoft/onnxruntime/pull/1959#discussion_r336695370) [](ancestors = 336695370)

File: onnxruntime/core/providers/cpu/ml/linearregressor.h (hunk 11 above)
Comment (snnn) [reply onnxruntime/core/providers/cpu/ml/linearregressor.h]: Will update this PR

File: onnxruntime/core/providers/cpu/ml/linearregressor.cc (hunk 7 above)
Comment (snnn) [reply onnxruntime/core/providers/cpu/ml/linearregressor.cc]: But, this is exactly how the old code works. If we change that, it's a breaking change.

Comment (snnn) [reply onnxruntime/core/providers/cpu/ml/linearregressor.cc]: I changed it to ORT_ENFORCE
